docs/docs/specs/v0.2/session-graph.md

```diff
@@ -0,0 +1,292 @@
+
+1. add content items to the UI,
+2. present content items in relation to other content items,
+3. present actions related to a content item,
+4. persist mutations to a content item in the UI,
+
+in such a way that the solution is:
+
+1. independent of any method of persistence beyond the UA’s own JS runtime,
+2. itself able to be persisted or replicated (including sharing),
+3. reasonably extensible by plugins without also causing corruptions.
+
+# Assumptions
+
+This proposal does not assume any particular limits on the host UA’s network bandwidth, memory, disk, or processing.
+
+# Proposed solution
+
+I propose `SessionGraph` should have an ontological structure similar to a relational database, where **Nodes** are stored separately from **Relations**.
+
+## Nodes
+
+Nodes must support the following:
+
+1. A scalar identifier unique within the session
+2. A static or translatable label
+3. A serializable, but possibly arbitrary, record of properties that a plugin can recognize and use as a way to resolve an entity the Node represents
+  - This _may_ include details about provenance
+  - This _may_ include a graphical way to present the node (references to an icon or image)
+
+## Relations
+
+Relations must support the following:
+
+1. One-to-one relation between one Node and another Node
+2. One-to-many relations between one Node and a set of Nodes
+
+## Not stored in `SessionGraph`: functions, class instances, etc
+
+So that the `SessionGraph` can be persisted and replicated, plugins must be able to resolve the entity a Node represents on their own at runtime.
+
+# Implementation
+
+In the rest of this proposal I will refer to the following types:
+
+```tsx
+type Identifier = string;
+
+type NodeKey = `dxos.org/${Identifier}`;
+
+export type Params =
+    | string
+    | number
+    | boolean
+    | null
+    | undefined
+    | Params[]
+    | Record<string, Params>;
+    
+type Node = {
+  id: NodeKey;
+  label: string | [string, {ns: string}];
+  params?: Record<string, Params>;
+};
+
+type RelationType = string;
+
+type RelationKey = `${RelationType}-${ 'child' | 'parent' }`;
+
+type SessionGraph = {
+  nodes: Deepsignal<
+    Map<NodeKey, Node>
+  >;
+  relations: Deepsignal<
+    Map<
+      NodeKey,
+      Map<
+        RelationKey,
+        NodeKey | Set<NodeKey>
+      >
+  >
+  >;
+};
+```
+
+## Examples for min-viable use-cases
+
+We agreed to several use-cases that any solution must facilitate.
+
+### Basics
+
+Plugins should be able to add content items and a node under which they can be organized by the plugin that added the content items.
+
+e.g.:
+> `MarkdownPlugin` can add nodes representing documents it finds and associate them with a node that represents itself as the documents’ provenance
+
+#### Adding content items
+
+```tsx
+const MARKDOWN_NS = 'dxos.org/markdown-plugin';
+
+const MARKDOWN_GROUP = `${MARKDOWN_PLUGIN}/documents`;
+
+const handleReady = async () => {
```

Comment:
Is this the general plugin ready callback or a graph-specific callback? Could this just be the graph `nodes` callback and return the set of nodes and relations?

Reply:
I left that to the reader to decide since I wasn't sure myself 😅 — wherever a plugin should produce nodes and relations from scratch, this is that function.

I will illustrate what plugins should do to keep the nodes updated.